### DISCUSSION TITLE
Monitor queue size with durable sessions enabled and community version

### DISCUSSION THREAD

**Discussion by GuillermoGarces** [emqx/emqx#14407] [Q&A] [answered]
I am using community version 5.8 with durable sessions enabled.
Dashboard does not inform clients queue size, and it says "Message list is unavailable when persistent session is enabled"
I tried also using the API  "​/clients​/{clientid}​/mqueue_messages" but it returns ""API not implemented for persistent sessions"

How do I know if messages are accumulating for a specific client? or how many messages are left for it to finish processing everything?

**Comment by ieQu1** [comment]
Hello,

> Dashboard does not inform clients queue size, and it says "Message list is unavailable when persistent session is enabled"

Durable sessions don't have the message queue, this is why this option is not available. They stream messages from a database.

> How do I know if messages are accumulating for a specific client?

Durable sessions don't accumulate messages, unlike memory sessions. When a durable session subscribes to a topic-filter, this topic-filter is marked as "durable". All new MQTT messages sent to such topics are saved into a builtin database which stores them on disk.
Durable sessions then "replay" messages stored in the database at their own pace. Therefore, they don't have to accumulate any messages. They only need to store *iterators*: objects that represent position of the last replayed message in the database for the given topic.

In reality it's more complicated than that, but the bottom line is that with this design the broker will never run out of RAM due to a lagging client, as clients don't accumulate messages in RAM.

> or how many messages are left for it to finish processing everything?

This is a valid concern. We are planning to add lag monitoring tools in the future.

**Comment by GuillermoGarces** [comment]
Thanks @ieQu1 

So durable sessions are handled as streams (like in kafka). All messages for a topic goes to the disk, they are persisted inside the builtin db and are kept there until they expire (retention period). Then all subscriber has an iterator (pointer) of the last processed message.

Current situation is then than there is nothing on the API or even by accessing the filesystem of the emqx server to get information on how many messages are in each stream/topic or which are the current values for each subscriber iterator
Is it like that?

What happens if I subscribe to a topic-filter whit a new client? Will it start its iterator pointing to the oldest message in the stream or the newest one?

Last question. Is there any way to connect and inspect the built-in database?

**Reply by ieQu1** [reply, marked as answer]
> So durable sessions are handled as streams (like in kafka). All messages for a topic goes to the disk, they are persisted inside the builtin db and are kept there until they expire (retention period). Then all subscriber has an iterator (pointer) of the last processed message.

Right. 

> Current situation is then than there is nothing on the API or even by accessing the filesystem of the emqx server to get information on how many messages are in each stream/topic or which are the current values for each subscriber iterator

Yes, right now there's no such API, but we're planning to add it. Monitoring database size is an important feature. Our current focus is on optimizing throughput of durable sessions, though.

Currently total disk utilization of the builtin DB can be monitored using `du -h --max-depth=0 data/messages/` (where data is a EMQX [data directory](https://docs.emqx.com/en/emqx/latest/deploy/install.html#files-and-directories))

> What happens if I subscribe to a topic-filter whit a new client? Will it start its iterator pointing to the oldest message in the stream or the newest one?

The iterators will point at the latest message in the stream. The new client won't receive any messages published before it subscribed to the topic, as per MQTT standard.

> Last question. Is there any way to connect and inspect the built-in database?

The builtin database provides an Erlang interface that caters to the durable session. 
You can login into the EMQX shell (`emqx remote_console`) and use 
[durable storage explorer module](https://github.com/emqx/emqx/blob/master/apps/emqx_durable_storage/src/emqx_dsx.erl) to inspect the messages. However, currently this module is only used internally for debugging, so it's not user-friendly by any means.
